pennylane_lightning_gpu/src/simulator/StateVectorCudaManaged.hpp

```diff
@@ -692,6 +692,146 @@ class StateVectorCudaManaged
         }
         return expval(obsName, wires, params, matrix_cu);
     }
+    /**
+     * @brief See `expval(std::vector<CFP_t> &gate_matrix = {})`
+     */
+    auto expval(const std::vector<size_t> &wires,
+                const std::vector<std::complex<Precision>> &gate_matrix) {
+
+        std::vector<CFP_t> matrix_cu(gate_matrix.size());
+
+        for (std::size_t i = 0; i < gate_matrix.size(); i++) {
+            matrix_cu[i] =
+                cuUtil::complexToCu<std::complex<Precision>>(gate_matrix[i]);
+        }
+
+        if (gate_matrix.empty()) {
+            std::string message = "Currently unsupported observable";
+            throw LightningException(message.c_str());
+        }
+
+        auto &&local_wires =
+            (gate_matrix.empty())
+                ? wires
+                : std::vector<size_t>{wires.rbegin(), wires.rend()};
+
+        auto expect_val =
+            getExpectationValueDeviceMatrix(matrix_cu.data(), local_wires);
+        return expect_val;
+    }
+
+    /**
+     * @brief Get expectation of a given host or device defined array.
+     *
+     * @param matrix Host or device defined row-major order gate matrix array.
+     * @param tgts Target qubits.
+     * @return auto Expectation value.
+     */
+    auto getExpectationValueOnSparseSpVM(
+        const int *csrOffsets_ptr, const int csrOffsets_size,
+        const int *columns_ptr, const std::complex<Precision> *values_ptr,
+        const int numNNZ) {
+
+        const int nIndexBits = BaseType::getNumQubits();
+        const int length = 1 << nIndexBits;
+        const int num_rows = csrOffsets_size - 1;
+        const int num_cols = num_rows;
+
+        int *d_csrOffsets, *d_columns;
+        CFP_t *d_values, *d_tmp;
+
+        Precision alpha = 1.0;
+        Precision beta = 0.0;
+
+        Precision expect = 0;
+
+        PL_CUDA_IS_SUCCESS(
+            cudaMalloc((void **)&d_csrOffsets, csrOffsets_size * sizeof(int)))
+        PL_CUDA_IS_SUCCESS(
+            cudaMalloc((void **)&d_columns, numNNZ * sizeof(int)))
+        PL_CUDA_IS_SUCCESS(
+            cudaMalloc((void **)&d_values, numNNZ * sizeof(CFP_t)))
+        PL_CUDA_IS_SUCCESS(cudaMalloc((void **)&d_tmp, length * sizeof(CFP_t)))
+
+        PL_CUDA_IS_SUCCESS(cudaMemcpy(d_csrOffsets, csrOffsets_ptr,
+                                      csrOffsets_size * sizeof(int),
+                                      cudaMemcpyHostToDevice))
+        PL_CUDA_IS_SUCCESS(cudaMemcpy(d_columns, columns_ptr,
+                                      numNNZ * sizeof(int),
+                                      cudaMemcpyHostToDevice))
+        PL_CUDA_IS_SUCCESS(cudaMemcpy(d_values, values_ptr,
+                                      numNNZ * sizeof(CFP_t),
+                                      cudaMemcpyHostToDevice))
+
```

Comment:
We already have a Cuda buffer datatype that abstracts away the memory allocations using CUDA calls. It would be best to avoid direct use of the `cudaMalloc` and `cudaMemcpy` where we can. See https://github.com/PennyLaneAI/pennylane-lightning-gpu/blob/main/pennylane_lightning_gpu/src/util/DataBuffer.hpp